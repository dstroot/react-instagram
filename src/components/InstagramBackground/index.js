import React, { useEffect, useState } from 'react';
import styled from 'styled-components';
import PropTypes from 'prop-types';

// get data
import { useQuery } from 'react-query';
import { fetchNetlify } from './queries';

// https://gist.github.com/kjintroverted/d67c7f12f68288f6ccf07cbd06fa66a8

const InstagramBackground = ({ username, quality }) => {
  const [images, setImages] = useState(null);
  const [imageDims, setImageDims] = useState(0);
  const { data, isLoading, error } = useQuery(
    ['photos', { username }],
    fetchNetlify,
    {
      // suspense: true,
      retry: 3,
    }
  );

  // loads images from data
  useEffect(() => {
    if (!error && data) {
      const { graphql } = data;
      setImages(
        graphql.user.edge_owner_to_timeline_media.edges.map(
          ({ node }) => node.thumbnail_resources
        )
      );
    }
  }, [data, error]);

  // calculate the width of the tiles
  // https://ryanve.com/lab/dimensions/
  function calcImageDims() {
    if (!images || !images.length) {
      setImageDims(0);
    } else {
      setImageDims(
        Math.floor(
          Math.sqrt((window.outerHeight * window.outerWidth) / images.length)
        )
      );
    }
  }

  // updates tile dimensions on image load
  useEffect(calcImageDims, [images]);

  // update tiles when window resizes (with cleanup)
  useEffect(() => {
    window.addEventListener('resize', calcImageDims);

    return function cleanupListener() {
      window.removeEventListener('resize', calcImageDims);
    };
  });

  /**
   * Styles
   */

  // PARENT TILE FOR IMAGES
  const Tile = styled.div`
    width: ${imageDims}px;
    height: ${imageDims}px;
    flex-grow: 1;
  `;

  const Post = styled.img`
    height: 100%;
    width: 100%;
    object-fit: cover;
  `;

  const Container = styled.div`
    position: fixed;
    width: 100vw;
    height: 100vh;
    display: flex;
    flex-wrap: wrap;
  `;

  if (!error && !isLoading) {
    return (
      <Container data-testid="container">
        {images &&
          images.map(res => (
            <Tile key={res[0].src}>
              <Post src={res[quality || 1].src} alt="recent post" />
            </Tile>
          ))}
      </Container>
    );
  }

  return null;
};

export default InstagramBackground;

InstagramBackground.propTypes = {
  /** The username of the Instagram account to use  */
  username: PropTypes.string.isRequired,
  /** The quality of the images. Range is 0-4. 0 = 150x150, 1 = 240x240, 2 = 320x320, 3 = 480x480, 4 = 640x640. Defaults to 1 */
  quality: PropTypes.number,
};

// Specifies the default values for props:
InstagramBackground.defaultProps = {
  quality: 1,
};
