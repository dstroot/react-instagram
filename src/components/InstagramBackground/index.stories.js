import React from 'react';
import styled from 'styled-components';

// components
import { InstagramBackground } from '.';

export default {
  title: 'Pac Life/Instagram Background',
  component: InstagramBackground,
};

export const example = () => {
  return (
    <Main>
      <InstagramBackground
        username="ferrytalecreative"
        filterOpts={['to bottom right', 'teal', 'blue', 'purple']}
      />
      <Fence>
        <Splash>
          <Title>Ferry Tale Creative</Title>
          <Lead>
            Thanks for visiting us! Follow us on Instagram to get the lastest
            looks. Also check out page background!
          </Lead>
          <Lead>
            Our store will soon be open for business. Until then look around and
            let us know what you think.
          </Lead>
          <Button>@ferrytalecreative</Button>
        </Splash>
        <Spacer />
      </Fence>
    </Main>
  );
};

const Main = styled.div`
  display: flex;
  justify-content: center;
`;

const Fence = styled.div`
  position: fixed;
  height: 100vh;
  width: 100%;
  max-width: 1000px;
  display: flex;
  align-items: center;
`;

const Spacer = styled.div`
  flex: 1;
`;

const Splash = styled.div`
  position: relative;
  left: 100px;
  width: 40vw;
  height: 50vh;
`;

const Title = styled.h1`
  font-size: 4em;
  color: white;
  font-weight: 300;
`;

const Lead = styled.p`
  font-size: 1.25em;
  color: white;
  font-weight: 300;
`;

const Button = styled.button`
  /* Adapt the colors based on primary prop */
  color: white;
  background-color: transparent;
  font-size: 1em;
  padding: 0.5em 1em;
  border: 2px solid white;
  border-radius: 3px;
`;
