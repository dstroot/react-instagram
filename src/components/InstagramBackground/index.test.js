import React, { Suspense } from 'react';
import { render, waitForElement, cleanup } from '@testing-library/react';

// component to test
import InstagramBackground from '.';
import data from './__mocks__/data.json';

// automatically unmount and cleanup DOM after each test is finished.
afterEach(cleanup);

describe('InstagramBackground', () => {
  // cleanup mocks
  beforeEach(() => {
    fetch.resetMocks();
  });

  it('it renders with suspense', async () => {
    fetch.mockResponseOnce(data);

    // const { getByTestId } = render(
    //   <Suspense fallback={<div>Not Rendered Yet</div>}>
    //     <InstagramBackground username="ferrytalecreative" />
    //   </Suspense>
    // );
    // const element = await waitForElement(() => getByTestId('container'));
    // expect(element.toMatchSnapshot());

    const { asFragment } = await waitForElement(() =>
      render(
        <Suspense fallback={<div>Not Rendered Yet</div>}>
          <InstagramBackground username="ferrytalecreative" />
        </Suspense>
      )
    );

    // assert the snapshot matches
    expect(asFragment()).toMatchSnapshot();

    //assert on the times called and arguments given to fetch
    expect(fetch.mock.calls.length).toEqual(1);
    expect(fetch.mock.calls[0][0]).toEqual(
      '/.netlify/functions/photos?username=ferrytalecreative'
    );
  });
});

/**
 * Custom matchers:
 *   - toBeDisabled
 *   - toBeEnabled
 *   - toBeEmpty
 *   - toBeInTheDocument
 *   - toBeInvalid
 *   - toBeRequired
 *   - toBeValid
 *   - toBeVisible
 *   - toContainElement
 *   - toContainHTML
 *   - toHaveAttribute
 *   - toHaveClass
 *   - toHaveFocus
 *   - toHaveFormValues
 *   - toHaveStyle
 *   - toHaveTextContent
 */

/**
  * Based on the Guiding Principles, your test should resemble how users
  * interact with your code (component, page, etc.) as much as possible.
  * With this in mind, we recommend this order of priority:

  * Queries Accessible to Everyone queries that reflect the experience
  * of visual/mouse users as well as those that use assistive technology
  *
  * **getByLabelText**: Only really good for form fields, but this is the
  * number one method a user finds those elements, so it should be your top preference.
  *
  * **getByPlaceholderText**: A placeholder is not a substitute for a label. But if
  * that's all you have, then it's better than alternatives.
  *
  * getByText: Not useful for forms, but this is the number 1 method a user
  * finds other elements (like buttons to click), so it should be your top
  * preference for non-form elements.
  *
  * getByDisplayValue: The current value of a form element can be useful when
  * navigating a page with filled-in values.
  *
  * Semantic Queries HTML5 and ARIA compliant selectors. Note that the user
  * experience of interacting with these attributes varies greatly across
  * browsers and assistive technology.
  *
  * **getByAltText**: If your element is one which supports alt text (img, area,
  * and input), then you can use this to find that element.
` *`
  * **getByTitle**: The title attribute is not consistently read by screenreaders,
  * and is not visible by default for sighted users.
  *
  * **getByRole**: This can be used to select dialog boxes and other
  * difficult-to-capture elements in a more semantic way
  *
  * Test IDs
  *
  * **getByTestId**: The user cannot see (or hear) these, so this is only
  * recommended for cases where you can't match by text or it doesn't make
  * sense (the text is dynamic).
  */
