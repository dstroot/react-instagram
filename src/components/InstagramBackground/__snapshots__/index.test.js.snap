// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InstagramBackground it renders with suspense 1`] = `
<DocumentFragment>
  <div>
    Not Rendered Yet
  </div>
</DocumentFragment>
`;
